app/scripts/controllers/metametrics.js

```diff
@@ -443,13 +443,30 @@ export default class MetaMetricsController {
     }
   }
 
+  // Method to track metrics event
+  // If an actionId is provided createEventFragment, finalizeEventFragment are used to track event,
+  // this is to achieve idempotent behaviour.
+  // If actionid is not provided this._trackEvent is used to track the event.
+  trackEvent(payload, options) {
+    if (options?.actionId) {
+      const fragment = this.createEventFragment({
+        ...payload,
+        ...options,
+        successEvent: payload.event,
+      });
+      this.finalizeEventFragment(fragment.id);
+    } else {
+      this._trackEvent(payload, options);
+    }
+  }
```

Comment:
Big thanks to @darkwing for his suggestion to add coverage for more scenario, I discovered that calling finalise immediately will result in fragment to be removed from state and thus duplicate fragment can then be created again.

@brad-decker: can you plz suggest ?

I am thinking we can probably leave `trackMetaMetricsEvent` function non idempotent, for a couple of reason:

1. It is hard to make this function uses 100% idempotent as it is called not just from UI but also from other controller methods where `actionId` is not available
2. Probability of duplicate events being logged is not very high

Reply:
Alternatively we can supply a very small timeout to the fragment making it automatically finalize the fragment on the next iteration of the timer/alarm in the metametrics controller. At worst this delays events being sent by a small amount. A delay is probably better than potentially duplicating events in edge cases.

Reply:
@brad-decker : I tried out this change and I found that a delay does not guarantees idempotent behaviour.

Alternatively, if I remove `finalizeEventFragment` I believe orphaned fragment will be finalised eventually.

Please suggest.

Reply:
Orphaned fragments are only finalized with a failure event therefore not what we want. When i said add a small timeout i meant in the fragment payload. How does that change idempotency?

Reply:
Hey @brad-decker : I added timeout here `fragment.timeout = 20000;` but it seems to have no effect. I noticed that `finalizeEventFragment` function is not using `timeout`, is that a bug.

Reply:
I'm not sure I follow what you're saying. If we don't finalize the event, nothing will ever be tracked. The snippet I posted is to explain why my original timeout idea won't work, we aren't supplying a "failureEvent" which is like an abandoned cart event if the user doesn't complete a fragment in a timeframe.

We might want to just move forward without making this action idempotent for now, the alternative path would be to add a mechanism for finalizing a fragment without the `abandoned: true` part after a timeout, or in metamask-controller's trackEvent wrapper:

1. Try to get an existing fragment by id
2. If it doesn't exist, create it
3. use a setTimeout to finalize the event after a small delay to ensure that identical actionId calls do not cause duplicate event tracking. 

Thoughts?

Reply:
yes